feat: new version metric (#13269)
argoproj/argo-workflows: docs/metrics.md

    Some metric attributes may have high cardinality and are marked with ⚠️ to warn you. You may need to disable this metric or disable the attribute.
<!-- titles should be the exact metric name for deep-linking, alphabetical ordered -->
<!-- titles are without argo_workflows prefix -->

> #### `controller_build_info`

Reviewer: Or just `version` to match `argo version`
```suggestion
#### `version`
```

> #### `controller_build_info`

Author: Version appears to match [otel semantic conventions](https://github.com/open-telemetry/semantic-conventions) use, so will switch to that.

The build information for this workflow controller

| attribute   | explanation                                                            |
|-------------|------------------------------------------------------------------------|
| `version`   | The version of argo-workflows                                          |
| `platform`  | Platform this is running on, as go describes it e.g. `linux/amd64` |

> | `gover`     | Version of go that built this workflow controller                  |

Reviewer: also isn't the convention to use underscores? or is it different for attribute names vs metric names?

```suggestion
| `go_version`     | Version of Go used                  |
```

> | `gover`     | Version of go that built this workflow controller                  |

Reviewer: Underscores are explicitly mentioned as word delimiters for attribute names in the convention docs: https://opentelemetry.io/docs/specs/semconv/general/attribute-naming/

So the second suggestion should be used.

It would also be be consistent with `argo version` which prints `GoVersion`

> | `gover`     | Version of go that built this workflow controller                  |

Reviewer: Small follow-up: https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1669064430

> | `build`     | Build date for this workflow controller                                |

Reviewer: Also it would be good to be consistent with `argo version` which prints `BuildDate`

> | `build`     | Build date for this workflow controller                                |

Reviewer: Follow-up: https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1669066371

> | `compiler`  | The compiler used to build this workflow controller e.g. `gc`          |

Reviewer: ```suggestion
| `compiler`  | The compiler used. Example: `gc`          |
```

same as above

Reviewer: Follow-up: https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1669066371

> | `commit`    | The full git SHA1 of the code that built this workflow controller      |
> | `treestate` | Whether the git tree was `dirty` or `clean` when this was built        |
> | `tag`       | The tag on the git commit or `untagged` if it was not tagged           |

Reviewer: ```suggestion
| `git_commit`    | The full Git SHA1 commit      |
| `git_tree_state` | Whether the Git tree was `dirty` or `clean` when built        |
| `git_tag`       | The Git tag or `untagged` if it was not tagged           |
```

- match `argo version` and split words with underscores per OTel convention
- simplify
- "Git" as proper noun

Author: Done

See [queue adds count](#queue_adds_count) for details.

#### `version`

Build metadata for this Controller.

| attribute        | explanation                                                                                           |
|------------------|-------------------------------------------------------------------------------------------------------|
| `version`        | The version of Argo                                                                                   |
| `platform`       | The [Go platform](https://go.dev/doc/install/source#environment) compiled for. Example: `linux/amd64` |

> | `goversion`      | Version of Go used                                                                                    |

Reviewer: ```suggestion
| `go_version`     | Version of Go used                                                                                    |
```

same as above and https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1661378969

> | `build_date`     | Build date for this workflow controller                                                               |
> | `compiler`       | The compiler used to build this workflow controller e.g. `gc`                                         |

Reviewer: ```suggestion
| `build_date`     | Build date                                                               |
| `compiler`       | The compiler used. Example: `gc`                                         |
```

You missed these two [previous](https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1661377037) [suggestions](https://github.com/argoproj/argo-workflows/pull/13269#discussion_r1661391935). GH collapsed those two, so I'm guessing you just didn't see it
